pennylane/templates/subroutines/all_singles_doubles.py

```diff
@@ -0,0 +1,180 @@
+    :math:`n`-qubit system in a superposition of the initial Hartree-Fock state and
+    other states encoding multiply-excited configurations.
+
+    Args:
+        weights (tensor_like): size ``(len(singles) + len(doubles),)`` tensor containing the
+            angles entering the :class:`~.pennylane.SingleExcitation` and
+            :class:`~.pennylane.DoubleExcitation` operations, in that order
+        wires (Iterable): wires that the template acts on
+        hf_state (array[int]): Length ``len(wires)`` occupation-number vector representing the
+            Hartree-Fock state. ``hf_state`` is used to initialize the wires.
+        singles (Sequence[Sequence]): sequence of lists with the indices of the two qubits
+            the :class:`~.pennylane.SingleExcitation` operations act on
+        doubles (Sequence[Sequence]): sequence of lists with the indices of the four qubits
+            the :class:`~.pennylane.DoubleExcitation` operations act on
+
+    .. UsageDetails::
+
+        Notice that:
+
+        #. The number of wires has to be equal to the number of spin orbitals included in
+           the active space.
+
+        #. The single and double excitations can be generated with the function
+           :func:`~.excitations`. See example below.
+
+        An example of how to use this template is shown below:
+
+        .. code-block:: python
+
+            import pennylane as qml
+            import numpy as np
+
+            electrons = 2
+            qubits = 4
+
+            # Define the HF state
+            hf_state = qml.qchem.hf_state(electrons, qubits)
+
+            # Generate all single and double excitations
+            singles, doubles = qml.qchem.excitations(electrons, qubits)
+
+            # Define the device
+            dev = qml.device('default.qubit', wires=qubits)
+
+            wires = range(qubits)
+
+            @qml.qnode(dev)
+            def circuit(weights, hf_state, singles, doubles):
+                qml.templates.AllSinglesDoubles(weights, wires, hf_state, singles, doubles)
+                return qml.expval(qml.PauliZ(0))
+
+            # Evaluate the QNode for a given set of parameters
+            params = np.random.normal(0, np.pi, len(singles) + len(doubles))
+            circuit(params, hf_state, singles=singles, doubles=doubles)
+    """
+
+    num_params = 1
+    num_wires = AnyWires
+    par_domain = "A"
+
+    def __init__(self, weights, wires, hf_state, singles=None, doubles=None, do_queue=True):
+
+        if len(wires) < 2:
+            raise ValueError(
+                "The number of qubits (wires) can not be less than 2; got len(wires) = {}".format(
+                    len(wires)
+                )
+            )
+
+        if not singles and not doubles:
+            raise ValueError(
+                "'singles' and 'doubles' lists can not be both empty; got singles = {}, doubles = {}".format(
+                    singles, doubles
+                )
+            )
```

Comment:
It's still a matter of taste, but we are trying to keep the checks to an absolute minimum, because they are used in every qnode call...

My thinking is that we should check tensor dimensions (since this is an annoying error to see), and things that are theoretically tricky. So if this check is something a user with a look at the docs can get right, maybe we should delete it?

Non-blocking comment though

Reply:
Good point. My concern in removing this check is that the user may want to call this template without passing ``singles`` and ``doubles``. In this case, the template will not raise an error, will only prepare the HF state and numbers will come out (even though the state the user is preparing is trivial). I vote to keep it :-).

Reply:
I didn't know that `not a` evaluates to True if `a=None`, interesting!